sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/multi/MultiServiceEventClient.java

```diff
@@ -0,0 +1,402 @@
+        }
+        deregisterStep.setComplete();
+    }
+
+    /**
+     * Forwards the provided offer(s) to all enclosed services, seeing which services are interested in them.
+     *
+     * TODO(data-agility): Lots of opportunities to optimize this. Needs benchmarks. For example:
+     * <ul>
+     * <li>- Hide reserved resources from services that they don't belong to</li>
+     * <li>- Forward the offers to a random ordering of services to avoid some services starving others</li>
+     * <li>- Distribute the offers across all services in parallel (optimistic offers)</li>
+     * <li>- ... Pretty much anything that you could see Mesos itself doing.</li>
+     * </ul>
+     */
+    @Override
+    public OfferResponse offers(Collection<Protos.Offer> offers) {
+        // If uninstalling, then finish uninstall. Otherwise decline short.
+        boolean noClients = false;
+        // Decline short.
+        boolean anyClientsNotReady = false;
+
+        List<OfferRecommendation> recommendations = new ArrayList<>();
+        List<Protos.Offer> remainingOffers = new ArrayList<>();
+        remainingOffers.addAll(offers);
+
+        Collection<String> finishedServices = new ArrayList<>();
+        Collection<String> uninstalledServices = new ArrayList<>();
+
+        Collection<AbstractScheduler> services = multiServiceManager.lockAndGetServices();
+        LOGGER.info("Sending {} offer{} to {} service{}:",
+                offers.size(), offers.size() == 1 ? "" : "s",
+                services.size(), services.size() == 1 ? "" : "s");
+        try {
+            if (services.isEmpty()) {
+                // If we don't have any clients, then WE aren't ready.
+                // Decline short, or finish if there's an uninstall in progress.
+                noClients = true;
+            }
+            for (AbstractScheduler service : services) {
+                String serviceName = service.getServiceSpec().getName();
+                OfferResponse response = service.offers(remainingOffers);
+                if (!remainingOffers.isEmpty() && !response.recommendations.isEmpty()) {
+                    // Some offers were consumed. Update what remains to offer to the next service.
+                    List<Protos.Offer> updatedRemainingOffers =
+                            OfferUtils.filterOutAcceptedOffers(remainingOffers, response.recommendations);
+                    remainingOffers = updatedRemainingOffers;
+                }
+                recommendations.addAll(response.recommendations);
+                LOGGER.info("  {} offer result: {}[{} rec{}], {} offer{} remaining",
+                        serviceName,
+                        response.result,
+                        response.recommendations.size(), response.recommendations.size() == 1 ? "" : "s",
+                        remainingOffers.size(), remainingOffers.size() == 1 ? "" : "s");
+
+                switch (response.result) {
```

Comment:
This is another case where combine service lifecycle and offer processing logic.  We should note this as a place where we might want to pursue de-coupling.

Reply:
(INFINITY-3493)